src/ir/auth_logic/universe_relation_insertion.h

```diff
@@ -0,0 +1,449 @@
+            prin,  // speaker
+            {Assertion(Fact({}, MakeUniverseMembershipFact(typing.first,
+                                                           typing.second)))}));
+      }
+    }
+    return universe_defining_says_assertions;
+  }
+
+  // This visitor populates a mapping (flat_hash_map) of constant (literal)
+  // names to their types by looking at the relations to which the literals
+  // are applied and declarations of those relations (including the types
+  // of the parameters where they are applied). It takes as input
+  // a mapping from relation names to relation declarations.
+  class TypeEnvironmentGenerationVisitor
+      : public AuthLogicAstTraversingVisitor<TypeEnvironmentGenerationVisitor> {
+   public:
+    // A map from the names of principal / predicate argument literals
+    // to typings.
+    using TypeEnv = absl::flat_hash_map<std::string, datalog::ArgumentType>;
+    // A map from predicate names to predicate declarations.
+    using DeclEnv =
+        absl::flat_hash_map<std::string, datalog::RelationDeclaration>;
+
+    TypeEnvironmentGenerationVisitor(
+        absl::flat_hash_map<std::string, datalog::RelationDeclaration> decl_env)
+        : decl_env_(decl_env) {}
+
+    const TypeEnv& type_env() const { return type_env_; }
+
+   private:
+    void AddTyping(std::string arg_name, datalog::ArgumentType arg_type) {
+      if (!IsNameConstant(arg_name)) {
+        // This pass is only meant to find typings for constants,
+        // so do nothing if this argument is not a constant.
+        return;
+      }
+      if (type_env_.find(arg_name) == type_env_.end()) {
+        // No typing for this constant was found yet,
+        // so add it.
+        type_env_.insert({arg_name, arg_type});
+      } else if (type_env_.find(arg_name)->second != arg_type) {
+        // If there is already a typing for this constant,
+        // but it is different from the one used here,
+        // throw a type error and abort.
+        LOG(FATAL) << "type error for constant: " << arg_name;
+      }
+    }
+
+    // (TODO: #672) once an AST node for numbers as RVALUEs in numeric
+    // comparisons is added, this visitor should also visit
+    // the numeric RVALUEs and add them to the type environment
+    // with type ArgumentType(kNumber, "Number")
+    // This is a workaround that makes use of the fact that
+    // numeric comparisons are represented as predicates
+    // with a name that matches the operator:
+    // (https://github.com/google-research/raksha/blob/be6ef8e1e1a20735a06637c12db9ed0b87e3d2a2/src/ir/auth_logic/ast_construction.cc#L92)
+    static inline bool PredicateIsNumericOperator(
+        const datalog::Predicate& pred) {
+      if (pred.name() == "<" || pred.name() == "<" || pred.name() == "=" ||
+          pred.name() == "!=" || pred.name() == "<=" || pred.name() == ">=") {
+        return true;
+      } else {
+        return false;
+      }
+    }
```

Comment:
Should this just be a member function on `datalog::Predicate`?

Reply:
fixed